File: src/tools/adapters/reactor/reactorTools.ts
/*
 * Copyright (c) 2025-2026 Datalayer, Inc.
 * Distributed under the terms of the Modified BSD License.
 */

/**
 * Reactor tool bundles, as tools the browser loop can call.
 *
 * A `ReactorToolSpec` names a plugin's commands and its backend. Both become
 * `FrontendToolDefinition`s here — the same shape the notebook tools have, so
 * a host adds them to an agent the same way — but they execute differently:
 *
 * - a **command** tool runs `reactor.executeCommand(command, args)` on the
 *   reactor the host is mounted in, which is exactly what the keystroke does;
 * - a **backend** tool makes the HTTP call the spec describes, from the page.
 *   On a server harness the same call is made by the Python side instead, see
 *   `agent_runtimes.services.reactor_tools`; a browser loop has no server, so
 *   it is embedded here.
 *
 * @module tools/adapters/reactor
 */

import { useMemo } from 'react';
import type { ReactorPlatform } from '@datalayer/reactor';
import { useOptionalReactorPlatform } from '@datalayer/reactor/react';
import type {
  FrontendToolDefinition,
  ReactorBackendToolSpec,
  ReactorCommandToolSpec,
  ReactorToolSpec,
} from '../../../types';

const NO_ARGUMENTS = { type: 'object', properties: {} } as const;

export type ReactorToolsOptions = {
  /** The reactor to execute commands on; `useReactorTools` finds the mounted one. */
  reactor?: ReactorPlatform | null;
  /** Where the backend is; the spec's `baseUrl`, then the page's origin, otherwise. */
  backendUrl?: string;
  /** For tests. */
  fetch?: typeof fetch;
};

/** Where a bundle's backend is reached from the page. */
export function resolveReactorBackendUrl(
  spec: ReactorToolSpec,
  override?: string,
): string {
  const base =
    override ??
    spec.backend?.baseUrl ??
    (typeof window !== 'undefined' ? window.location.origin : '');
  return base.replace(/\/+$/, '');
}

/**
 * The request a backend tool makes for a set of arguments.
 *
 * Properties named in the path — `/decks/{id}` — are substituted; the rest
 * are query parameters for GET and DELETE and the JSON body otherwise. Kept
 * separate from the fetch so it can be checked without a network.
 */
export function buildReactorBackendRequest(
  tool: ReactorBackendToolSpec,
  baseUrl: string,
  args: Record<string, unknown>,
): { url: string; init: RequestInit } {
  const remaining: Record<string, unknown> = { ...args };
  const path = tool.path.replace(/\{(\w+)\}/g, (_, name: string) => {
    const value = remaining[name];
    delete remaining[name];
    return value === undefined || value === null
      ? ''
      : encodeURIComponent(String(value)).replace(/%2F/g, '/');
  });
  const method = tool.method.toUpperCase();
  const init: RequestInit = { method, headers: { accept: 'application/json' } };
  let url = `${baseUrl}${path}`;
  if (method === 'GET' || method === 'DELETE') {
    const query = new URLSearchParams();
    for (const [key, value] of Object.entries(remaining)) {
      if (value !== undefined && value !== null) {
        query.set(
          key,
          typeof value === 'object' ? JSON.stringify(value) : String(value),
        );
      }
    }
    const search = query.toString();
    if (search) {
      url += `?${search}`;
    }
  } else {
    init.headers = { ...init.headers, 'content-type': 'application/json' };
    init.body = JSON.stringify(remaining);
  }
  return { url, init };
}

/** The bundle's commands, executed on a reactor. */
export function reactorCommandTools(
  spec: ReactorToolSpec,
  reactor: ReactorPlatform | null | undefined,
): FrontendToolDefinition[] {
  return spec.frontend.map((entry: ReactorCommandToolSpec) => ({
    name: entry.name,
    description: entry.description,
    parameters: entry.parameters ?? NO_ARGUMENTS,
    location: 'frontend' as const,
    handler: async (args: Record<string, unknown>) => {
      if (!reactor) {
        throw new Error(
          `${entry.name}: the ${spec.plugin ?? spec.id} plugin is not mounted on this page.`,
        );
      }
      const argument = Object.keys(args ?? {}).length > 0 ? args : undefined;
      await reactor.executeCommand(entry.command, argument);
      return { ok: true, command: entry.command, argument };
    },
  }));
}

/** The bundle's backend, called from the page. */
export function reactorBackendTools(
  spec: ReactorToolSpec,
  options: Pick<ReactorToolsOptions, 'backendUrl' | 'fetch'> = {},
): FrontendToolDefinition[] {
  const backend = spec.backend;
  if (!backend) {
    return [];
  }
  const baseUrl = resolveReactorBackendUrl(spec, options.backendUrl);
  const doFetch =
    options.fetch ?? (typeof fetch === 'function' ? fetch : undefined);
  return backend.tools.map((tool: ReactorBackendToolSpec) => ({
    name: tool.name,
    description: tool.description,
    parameters: tool.parameters ?? NO_ARGUMENTS,
    location: 'frontend' as const,
    handler: async (args: Record<string, unknown>) => {
      if (!doFetch) {
        throw new Error(
          `${tool.name}: no fetch available to reach ${baseUrl}.`,
        );
      }
      const { url, init } = buildReactorBackendRequest(
        tool,
        baseUrl,
        args ?? {},
      );
      const response = await doFetch(url, init);
      if (!response.ok) {
        const detail = await response.text().catch(() => '');
        throw new Error(
          `${tool.name}: ${tool.method} ${url} answered ${response.status}${detail ? ` — ${detail}` : ''}`,
        );
      }
      if (response.status === 204) {
        return { ok: true };
      }
      const text = await response.text();
      try {
        return text ? JSON.parse(text) : { ok: true };
      } catch {
        return text;
      }
    },
  }));
}

/** Every tool of every bundle: commands first, then the backend. */
export function reactorToolDefinitions(
  specs: ReactorToolSpec[],
  options: ReactorToolsOptions = {},
): FrontendToolDefinition[] {
  return specs
    .filter(spec => spec.enabled !== false)
    .flatMap(spec => [
      ...reactorCommandTools(spec, options.reactor),
      ...reactorBackendTools(spec, options),
    ]);
}

/**
 * The tools of an agent's `reactorTools`, bound to the reactor mounted on
 * this page.
 *
 * Tolerant of there being none: the command tools then fail with a message
 * naming the plugin, which is a better answer to the model than a missing
 * tool, and the backend tools work regardless.
 */
export function useReactorTools(
  specs: ReactorToolSpec[] | undefined,
  options: Omit<ReactorToolsOptions, 'reactor'> = {},
): FrontendToolDefinition[] {
  const reactor = useOptionalReactorPlatform();
  const { backendUrl, fetch: fetchImpl } = options;
  return useMemo(
    () =>
      specs && specs.length > 0
        ? reactorToolDefinitions(specs, {
            reactor,
            backendUrl,
            fetch: fetchImpl,
          })
        : [],
    [specs, reactor, backendUrl, fetchImpl],
  );
}

File: agent_runtimes/services/reactor_tools.py
# Copyright (c) 2025-2026 Datalayer, Inc.
# Distributed under the terms of the Modified BSD License.

"""Reactor tool bundles on the server: the backend half, as pydantic-ai tools.

A ``ReactorToolSpec`` has two halves. Its *frontend* entries are reactor
commands and can only run where the plugin is mounted — the browser sends
those as client tools with every request, the way it sends the notebook
tools, and executes them itself when the model calls one. Its *backend*
entries are the plugin's HTTP API, and those the harness can call directly:
this module turns them into a :class:`FunctionToolset` an agent runs with,
whichever transport (AG-UI, Vercel AI) is in front of it.

Each tool is built from the JSON Schema in the spec with
:meth:`pydantic_ai.Tool.from_schema`, so the model sees exactly the arguments
the spec declared, and the call is one HTTP request: path parameters
substituted, the rest as query for GET and DELETE and as the JSON body
otherwise — the same rule the browser adapter follows.
"""

from __future__ import annotations

import logging
import os
import re
from typing import Any, Callable, Iterable, List, Optional
from urllib.parse import quote

import httpx

from agent_runtimes.specs.reactor_tools import get_reactor_tool_spec
from agent_runtimes.types import ReactorBackendToolSpec, ReactorToolSpec

logger = logging.getLogger(__name__)

_PATH_PARAM = re.compile(r"\{(\w+)\}")
_NO_ARGUMENTS: dict[str, Any] = {"type": "object", "properties": {}}


def resolve_reactor_backend_url(
    spec: ReactorToolSpec, override: Optional[str] = None
) -> Optional[str]:
    """Where the bundle's backend is: an override, its env var, its default."""
    backend = spec.backend
    if backend is None:
        return None
    candidates = [override]
    if backend.base_url_envvar:
        candidates.append(os.environ.get(backend.base_url_envvar))
    candidates.append(backend.base_url)
    for candidate in candidates:
        if candidate:
            return candidate.rstrip("/")
    return None


def build_reactor_backend_request(
    tool: ReactorBackendToolSpec, base_url: str, args: dict[str, Any]
) -> tuple[str, str, dict[str, Any], Optional[dict[str, Any]]]:
    """``(method, url, query, json_body)`` for a call of ``tool`` with ``args``."""
    remaining = dict(args)

    def substitute(match: re.Match[str]) -> str:
        value = remaining.pop(match.group(1), None)
        return "" if value is None else quote(str(value), safe="/")

    path = _PATH_PARAM.sub(substitute, tool.path)
    method = tool.method.upper()
    url = f"{base_url}{path}"
    if method in ("GET", "DELETE"):
        query = {
            key: value for key, value in remaining.items() if value is not None
        }
        return method, url, query, None
    return method, url, {}, remaining


def _make_tool_function(
    tool: ReactorBackendToolSpec,
    base_url: str,
    client_factory: Callable[[], httpx.AsyncClient],
) -> Callable[..., Any]:
    async def call(**kwargs: Any) -> Any:
        method, url, query, body = build_reactor_backend_request(tool, base_url, kwargs)
        async with client_factory() as client:
            response = await client.request(method, url, params=query or None, json=body)
        if response.status_code >= 400:
            # The model reads this. Say what was asked and what came back.
            return {
                "error": f"{method} {url} answered {response.status_code}",
                "detail": response.text[:2000],
            }
        if response.status_code == 204 or not response.content:
            return {"ok": True}
        try:
            return response.json()
        except ValueError:
            return response.text

    call.__name__ = tool.name
    call.__doc__ = tool.description
    return call


def reactor_backend_tools(
    reactor_tool_ids: Iterable[str],
    *,
    base_url: Optional[str] = None,
    client_factory: Optional[Callable[[], httpx.AsyncClient]] = None,
) -> list[Any]:
    """The backend tools of the named bundles, as ``pydantic_ai.Tool`` objects."""
    from pydantic_ai import Tool

    factory = client_factory or (lambda: httpx.AsyncClient(timeout=30.0))
    tools: list[Any] = []
    for tool_id in reactor_tool_ids:
        spec = get_reactor_tool_spec(tool_id)
        if spec is None:
            logger.warning("Reactor tool bundle '%s' not found; skipping", tool_id)
            continue
        if not spec.enabled or spec.backend is None:
            continue
        resolved = resolve_reactor_backend_url(spec, base_url)
        if not resolved:
            logger.warning(
                "Reactor tool bundle '%s' names no backend URL; skipping its backend tools",
                tool_id,
            )
            continue
        for entry in spec.backend.tools:
            tools.append(
                Tool.from_schema(
                    _make_tool_function(entry, resolved, factory),
                    name=entry.name,
                    description=entry.description or entry.name,
                    json_schema=entry.parameters or _NO_ARGUMENTS,
                    takes_ctx=False,
                )
            )
    return tools


def reactor_backend_toolset(
    reactor_tool_ids: Iterable[str],
    *,
    base_url: Optional[str] = None,
    client_factory: Optional[Callable[[], httpx.AsyncClient]] = None,
) -> Any | None:
    """A ``FunctionToolset`` of the bundles' backend tools, or ``None`` if there are none."""
    tools = reactor_backend_tools(
        reactor_tool_ids, base_url=base_url, client_factory=client_factory
    )
    if not tools:
        return None
    from pydantic_ai.toolsets import FunctionToolset

    return FunctionToolset(tools, id="reactor-tools")


def reactor_tool_names(reactor_tool_ids: Iterable[str]) -> dict[str, list[str]]:
    """``{"frontend": [...], "backend": [...]}`` — the tool names the bundles grant."""
    frontend: list[str] = []
    backend: list[str] = []
    for tool_id in reactor_tool_ids:
        spec = get_reactor_tool_spec(tool_id)
        if spec is None or not spec.enabled:
            continue
        frontend.extend(entry.name for entry in spec.frontend)
        if spec.backend:
            backend.extend(entry.name for entry in spec.backend.tools)
    return {"frontend": frontend, "backend": backend}


def reactor_tools_requiring_approval(reactor_tool_ids: Iterable[str]) -> List[str]:
    """Tool names whose spec asks for a person's approval before they run."""
    required: list[str] = []
    for tool_id in reactor_tool_ids:
        spec = get_reactor_tool_spec(tool_id)
        if spec is None or not spec.enabled:
            continue
        required.extend(e.name for e in spec.frontend if e.approval == "manual")
        if spec.backend:
            required.extend(e.name for e in spec.backend.tools if e.approval == "manual")
    return required
